### ISSUE TITLE
Two successive push not working

### ISSUE THREAD

**Issue by pcesari** [nodegit/nodegit#1320]
Hello,

Sorry for passing through the issues but the libgit2's Slack website is blocked by our company policy.

I have the following issue: I can not manage to have successive commit and push working in a row from a single html page.
We have developed an application using Electron with the following structure:
- one window used as GUI where the end user manipulates the data (layouts, calendars...)
- one window acting as a an abstract datalayer, using a dedicated npm module > this datalayer is requested through exposed API conceived methods which can be called from anywhere, and then stores data in json files in a dedicated folder acting as an internal database

We want to centralize data of users on our git using branches, therefore we have integrated some of nodegit functionalities in the datalayer page.
We have the problem that we can make the first commit and push correctly, but the second one does not fully work: the commit is done but the push operation not, and probably triggers an error which we can not see as the DevTools of Chrome crashes.
To try to isolate the issue of possible asynchronous errors, we have put all the code in a single function (here joint) where we make locally all the declarations using local variables: open the repository, build the tree from all paths of files in the repository, create the commit on HEAD, instantiate the remote, and finally push.

If, between the two calls to the commit and push operations in the datalayer page, we reload this page: the problem does not occur.

Can you think of a reason why the second time the push operation does not work?

**Comment by rcjsuen** [comment]
Does Electron itself crash or only the DevTools?

WIthout any code, it will be very difficult to make any claims as to why your application is not running. You may want to add debugging statements in NodeGit and in libgit2 and/or attach a C/C++ debugger to the Electron process.

**Comment by pcesari** [comment]
`<html>
<script type="text/javascript">

var NodeGit = require("nodegit");

var pathToRepo = "path_to_the_local_repo";
var cred = NodeGit.Cred.userpassPlaintextNew("my_login", "my_pwd");
var path_heads = "refs/heads/";
var login = "my_login";
var url = "url_to_central_repo.git";

function commit_test() {
	var repo_local;
	var remote_local;
	var treeOid_local;
	var index_local;
	var now = Math.round(Date.now() / 1000);
	var signature_local = NodeGit.Signature.create(login, "my_email", now, 0); 
	return NodeGit.Repository.open(pathToRepo)
	.then(function(repoResult) { 
					
		repo_local = repoResult;
		
		return repoResult.index()
	})
    .then(function(indexResult) {
      index_local = indexResult;
      index_local.read(1);
      var paths = [];

      return NodeGit.Status.foreach(repo_local, function(path) {
        paths.push(path);
      }).then(function() {
        return Promise.resolve(paths);
      });
    })
	.then(function(paths) {
	  return index_local.addAll(paths);
	})
	.then(function() {
	  return index_local.write();
	})
	.then(function(){
	  return index_local.writeTree();
	})
	.then(function(treeOidResult) {
	  treeOid_local = treeOidResult;
	return repo_local.getHeadCommit();
	})
	.then(function(parent) {
	  return repo_local.createCommit(
		'HEAD',
		signature_local,
		signature_local,
		("commit test push ").concat(Date.now().toString()),
		treeOid_local,
		[parent]);
	})
	.then(function() {
		return repo_local.getRemote("origin")
	})

	.then(function(remoteResult){
		remote_local = remoteResult;
	})
	.then(function() {
		return remote_local.push( 
		[path_heads.concat(login,":",path_heads,login)], 
			{ 
			  callbacks: {
				credentials: function() {
				  return cred;
				 },
				 transferProgress: function(stats) {
					 console.log("transfer progress");
				 }
			  }
		}
		) 
		
	})
	.catch(function(err) { console.log(err); });
}

// When a request is sent to this page and requires to save data, call the commit and push function
// This_page.onEvent{
	commit_test()
	.then(function(){
		console.log("finish commit and push, leave the hand to the page");
	})
	.catch(function(err) { throw err; });
//}
	
</script>
</html>
`

**Comment by pcesari** [comment]
@rcjusen : Only the DevTools crashes, when the push is being executed.

**Comment by rcjsuen** [comment]
@pcesari [Your code](https://github.com/nodegit/nodegit/issues/1320#issuecomment-314076582) is very difficult to read. Please edit your comment and reformat it.

**Comment by rcjsuen** [comment]
What if you try to run the webpage in Chrome instead of Electron?

**Comment by pcesari** [comment]
The problem is that our application in Electron is specifically developed around Electron, so it is hardly feasible to package it for use in Chrome, therefore I can not answer unfortunately.

**Comment by rcjsuen** [comment]
You say the commit completes but the push does not. Have you tried writing to a log file instead of using `console.log` since DevTools is crashing and you can't monitor things in the DevTools console?

**Comment by pcesari** [comment]
We managed to have successive push working, not sure exactly what was wrong though as we tried several ways of creating and setting up the branches...

**Comment by pcesari** [comment]
Thank you for your help. It would be nice to have more documentation on the API though, it is difficult to have the library working properly.

**Comment by rcjsuen** [comment]
@pcesari For the most part, NodeGit wraps libgit2's API. Browsing [their documentation](https://libgit2.github.com/libgit2/#HEAD) will help you get a better understanding of what functions do and what parameters to provide to the function.

**Comment by DMXL** [comment]
@pcesari Having the same issue here (in node.js environment though). Any idea how it was fixed or open sourced examples to look at?

**Comment by Unimax** [comment]
this issue is because of the cred object in the callback.
cannot reuse for some reason.

try return Git.Cred.userpassPlaintextNew("username", "password");
or something similar.
